phobos/core/robot.py

```diff
@@ -2622,12 +2627,60 @@ def derive_model_dictionary(root, name='', objectlist=[]):
         model['chains'][chain['name']] = chain
 
     # gather information on lights
-    # log("Parsing lights...", "INFO")
+    log.info("Parsing lights...")
     for obj in objectlist:
         if obj.phobostype == 'light':
             model['lights'][nUtils.getObjectName(obj)] = deriveLight(obj)
 
     # gather submechanism information from links
-    # log("Parsing submechanisms...", "INFO")
+    log.info("Parsing submechanisms...")
+
+    def getSubmechanisms(link):
+        """
+
+        Args:
+          link:
+
+        Returns:
 
+        """
+
+        if 'submechanism/name' in link.keys():
+            submech = {
+                'type': link['submechanism/type'],
+                'contextual_name': link['submechanism/name'],
+                'name': link['submechanism/subtype']
+                if 'submechanism/subtype' in link
+                else link['submechanism/type'],
+                'jointnames_independent': [
+                    nUtils.getObjectName(j, 'joint') for j in link['submechanism/independent']
+                ],
+                'jointnames_spanningtree': [
+                    nUtils.getObjectName(j, 'joint') for j in link['submechanism/spanningtree']
+                ],
+                'jointnames_active': [
+                    nUtils.getObjectName(j, 'joint') for j in link['submechanism/active']
+                ],
+                # TODO: this should work in almost all cases, still a bit of a hack:
+                'file_path': '../submechanisms/urdf/' + link['submechanism/name'] + '.urdf',
+            }
+            log.debug('    ' + submech['contextual_name'])
+        else:
+            submech = None
+        mechanisms = [submech] if submech else []
+        for c in link.children:
+            if c.phobostype in ['link', 'interface'] and c in objectlist:
+                mechanisms.extend(getSubmechanisms(c))
+        return mechanisms
+
+    model['submechanisms'] = getSubmechanisms(root)
+
+    # add additional data to model
+    model.update(deriveTextData(model['name']))
+
+    # shorten numbers in dictionary to n decimalPlaces and return it
+    log.info("Rounding numbers to {} digits.".format(ioUtils.getExpSettings().decimalPlaces))
+    model = roundFloatsInDict(model, ioUtils.getExpSettings().decimalPlaces)
+    log.debug("Sorting objects.")
+    model = sortListsInDict(model)
```

Comment:
Cool!!! Is this already tested?

Reply:
Can you split the MR into two? Then we can already merge the already working features seperately

Reply:
That function was previously used in blender-phobos and should in principle work. The entire get_robot_from_blender_dict needs a revision after last merge